pkg/sql/opt/optbuilder/mutation_builder.go

```diff
@@ -795,7 +795,22 @@ func (mb *mutationBuilder) addSynthesizedDefaultCols(
 }
 
 // addSynthesizedComputedCols is a helper method for addSynthesizedColsForInsert
-// and addSynthesizedColsForUpdate that scans the list of table columns, looking
+// and addSynthesizedColsForUpdate that projects computed column expressions and
+// adds the corresponding target table columns to targetColList and
+// targetColSet. See projectSynthesizedComputedCols for more details.
+//
+// NOTE: colIDs is updated with the column IDs of any synthesized columns which
+// are added to mb.outScope. If restrict is true, only columns that depend on
+// columns that were already in the list (plus all write-only columns) are
+// updated.
+func (mb *mutationBuilder) addSynthesizedComputedCols(colIDs opt.OptionalColList, restrict bool) {
+	mb.targetColSet = mb.projectSynthesizedComputedCols(colIDs, restrict)
+	for col, ok := mb.targetColSet.Next(0); ok; col, ok = mb.targetColSet.Next(col + 1) {
+		mb.targetColList = append(mb.targetColList, col)
```

Comment:
I think this assumes that table columns have ascending IDs... which is true right now, but maybe not something we want to rely on.

Here's one place where we assume that the order of columns in `targetColList` is the same as the order of columns in `mb.outScope` after `projectSynthesizedComputedCols` returns: https://github.com/cockroachdb/cockroach/blob/b845bd8573dd8787becad6df17424df2772c3940/pkg/sql/opt/optbuilder/insert.go#L699-L701

The need for `targetColList` is dubious... Maybe we can get rid of it in the near future...

In the meantime I'm going to attempt a safer fix.